### ISSUE TITLE
InvalidCastException: Unable to cast object of type 'CompositeText' to type 'ChangedText

### ISSUE THREAD

**Issue by JoshVarty** [dotnet/roslyn#10452]
**Version Used**: 
Visual Studio Update 2. (v14.0.25123.00)

**Steps to Reproduce**:
I don't have a minimum repro. I'm building a workspace that sits within Visual Studio. I convert [`INormalizedTextChangeCollection`](https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.text.inormalizedtextchangecollection.aspx) into [`TextChange`](http://source.roslyn.io/#Microsoft.CodeAnalysis/Text/TextChange.cs,9526992299f4ac04). Then I apply the changes to a `Document` and then apply the changes to the workspace via `Workspace.TryApplyChanges()`

**Expected Behavior**:

The updated document is applied to the workspace.

**Actual Behavior**:

Occasionally (seemingly with large changes, or quick changes) I receive the following exception.

```
[Error, Diagnostics v1.4.3-vs14.0-debug] System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.Text.CompositeText' to type 'Microsoft.CodeAnalysis.Text.ChangedText'.
   at Microsoft.CodeAnalysis.Text.ChangedText.WithChanges(IEnumerable`1 changes)
   at Microsoft.CodeAnalysis.Workspace.ApplyProjectChanges(ProjectChanges projectChanges)
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution)
   at CodeConnect.ExecutionEngine.SolutionTracker.applyChangesToDocument(String filePath, INormalizedTextChangeCollection changes) in C:\Users\Student\Documents\GitHub\Orange\src\ProjectOrange\CodeConnect.ExecutionEngine\SolutionTracker.cs:line 247
```

The line throwing the exception is: https://github.com/dotnet/roslyn/blob/6934bcb5fe8ec871b2f3c6da41a2ce1520facdc2/src/Compilers/Core/Portable/Text/ChangedText.cs#L86

It was introduced in: https://github.com/dotnet/roslyn/pull/7219 commit 6b726fbdf65f57bc5636cf15c00fdda6c0abb435

**Comment by weltkante** [comment]
This is reproducable and happened for me when applying no-op CompletionService changes (because the user already typed the whole word himself and then selected the completion). See issue #12216 (I've closed it because it is a duplicate of this one, which I didn't see at first).

Simplified code to reproduce the underlying issue:
```csharp
SourceText.From("Hello World")
    .WithChanges(new TextChange(new TextSpan(5, 6), string.Empty)) // prepare a ChangedText instance
    .WithChanges(); // this should be a no-op but currently throws an exception

SourceText.From("Hello World")
    .WithChanges(new TextChange(new TextSpan(5, 6), string.Empty)) // prepare a ChangedText instance
    .WithChanges(new TextChange(new TextSpan(2, 0), string.Empty)); // this should be a no-op but currently throws an exception
```

**Comment by weltkante** [comment]
@mattwar @Pilchie The bug is that `ChangedText.WithChanges` implementation assumes that the return value of arbitrary `SourceText.WithChanges` is always another ChangedText but this is not always true, for example no-op changes will return the original SourceText and trigger an InvalidCastException when the original SourceText was a CompositeText.

Since `SourceText.WithChanges` can be overridden the implementation in ChangedText should not assume any specific return value (even if it wasn't a no-op) or otherwise it should be documented on SourceText.WithChanges that subclass overrides have to respect specific assumptions about what to return.

**Comment by weltkante** [comment]
My above concern of `SourceText.WithChanges` being overridden has been answered in the discussion on PR #12230 , I'm quoting it here in case someone comes across the issue later.

> There are only a few types of SourceText that end up in _newText, CompositeText, SubText, LargeText or StringText. Though it does seem possible that if the change has only one segment and that segment is the entirety of the original text, then _newText will be the original text. So if that original text is a special subtype of SourceText that overrides the WithChanges API, it could return something that is not a ChangedText instance.

> Though in order to get into this particular situation, that subtype would have to have previously returned a ChangedText instance. And yet, ChangedText is internal, so custom subtypes could never return it.

> TL;DR - no, there is no concern, because it can't logically happen.
